### ISSUE TITLE
Can't deselect specific nodes with a second CMD click

### ISSUE THREAD

**Issue by nghiemt** [neo4j-apps/neo4j-bloom#60]
After Selecting multiple nodes using click and CMD Click (on the Mac), I can't deselect any of those nodes with a second CMD Click on the node. The only option that remains is to click elsewhere to deselect ALL selected nodes.

Using Bloom 2.8.1 on the Mac

**Comment by GlacialBoy** [comment]
Thanks @nghiemt for raising this topic, we will treat it as a feature request and it will be evaluated in the future releases of Bloom!

**Comment by ghanique** [comment]
@GlacialBoy I'd like to upvote this feature request.

Neo4j Bloom is pretty awesome. I use it extensively to analyze data, find correlations for data enhancement, etc. Regularly I find myself in the situation where I'm exploring a specific node, expanding specific relationships to other nodes. When I come to the conclusion that a specific approach is a dead end, I'd like to select the relevant node, press Ctrl-Shift-R to select related nodes, unselect the current node and then press Ctrl-H to delete all related nodes except the current one. Because I cannot remove the current node from the selection, this approach doesn't work, resulting in a lot of clicking and selecting to remove the irrelevant nodes some other way.

It gets worse when trying to subtract sets of nodes. For example, nodes A could relate to nodes B, which could relate to nodes C. Selecting node A, then Expand All, and Expand All one more time gives everything three levels deep. Now suppose not all B nodes have C nodes and that I want to examine those more closely. What would want do is just search for A-any-B-any-C and remove those with the exception of A. This, however, is not possible. A workaround would be to create a Search Phrase that returns (A)-[]-(B) where not (B)-[]-(C), but that is considerable more work.

In short, the inability to remove a node from a selection is a limitation I run into on a daily basis.

Thank you.

**Comment by GlacialBoy** [comment]
Thank @ghanique for the comment, this issue is quite an old one and it has been resolved over time.
It is possible to select and deselect using CMD+Click in MAC (or CTRL+Click in Win), the same also is valid if we use the Box Selector and the Lasso Selector.

Following in this video you can see:
1. Selection and Deselection with CMD+Click
2. Selection with Box and Deselection with CMD+Click
3. Selection with Lasso and Deselection with CMD+Click

https://github.com/user-attachments/assets/5d883ab1-b8e5-4658-bcb6-13511943d03a

We encourage you to use the latest version of Bloom that can be downloaded by the [neo4j donwload center](https://neo4j.com/deployment-center/#tools-tab) or directly using https://bloom.neo4j.io/
With both you can connect to a local database for free.

Notice that Bloom in Neo4j Desktop is discontinued and it will remain as version Bloom 2.11.0 (from November 2023)

Please let us now if using the latest version of Bloom is working for you or if you want to provide a video or more specific steps to reproduce the problem.

thanks!

**Comment by ghanique** [comment]
Hi @GlacialBoy ,

Thank you for your elaborate reply.

I love the video, and that is what I'm looking for. All I can say, then, is that it isn't working on Windows, not with the left control, nor the right, nor alt, nor shift... I can select nodes with the Box Selector, but holding Control and using the Box Selector again, doesn't unselect them.

I didn't know there was a Lasso Selector. Which makes me thinking...

If I hold the left mouse button and drag it immediately, it pans the screen (as if the mouse cursor is stuck to the background and moving the mouse moves the background). The only way that I know of to get a box selector is by holding the left mouse button, waiting a second, and then if I move the mouse I get a Box Selector.

Could it be that in Windows things are implemented differently, which is preventing the Control-Click from working properly?

Now I'm also noticing that you have icons in your video in the upper-left corner that are missing in mine.

<img width="173" alt="Image" src="https://github.com/user-attachments/assets/aa9a903e-cf25-4473-86be-90cf8ef92590" />

My version is 2.11.0, which according to your post should be the latest version.

**Comment by ghanique** [comment]
Some small update:

I noticed the Neo4j Download Center has a version 2.20.0. I downloaded the zip-file, but I couldn't make it work.

The zip file contains:

- 3 bloom-plugins.jar, one for 4.x, one for 5.x, one for 6.x
- bloom-2.20.0-assets.zip

Nowhere in my Neo4j Desktop installation can I find a bloom*.jar that I could replace with one of the jars.

There is a directory C:\Users\username\.Neo4jDesktop\neo4j-bloom:

<img width="125" alt="Image" src="https://github.com/user-attachments/assets/9e3cba22-08cb-4f31-bf5f-2d6e23a19e52" />

`dist` contains a directory `assets`, which is similar to `bloom-2.20.0-assets.zip`, which also contains an `assets` directory. I tried replacing everything in the dist folder with the contents of the zip-file, but that didn't work:

When I restart Neo4j Desktop, this is what "Bloom" looks like:

<img width="283" alt="Image" src="https://github.com/user-attachments/assets/6fd103c6-f1e1-41bd-8a4f-863e8c82c3ab" />

If I then try to open it:

<img width="437" alt="Image" src="https://github.com/user-attachments/assets/4594ddba-81bf-42ed-a1d5-7f9ec433681d" />

... I get:

<img width="377" alt="Image" src="https://github.com/user-attachments/assets/0f9ace51-71af-4798-88a0-0e1b76a7f887" />

Is there any way I can upgrade Neo4j Bloom in Neo4j Desktop?

(Or can a future version of Neo4j Desktop be upgraded to include Bloom 2.20.0?)

**Comment by ckanz** [comment]
Hi @ghanique,

the Bloom assets you can download in the Download Center aren't "[graph apps](https://neo4j.com/docs/desktop-manual/current/visual-tour/#graph-apps)", so it's not possible to install them in Neo4j Desktop this way. The Bloom graph app has been discontinued since Nov 2023. So unfortunately, the fix for issue you're describing in your comment https://github.com/neo4j-devtools/neo4j-bloom/issues/60#issue-1805927390 won't be available in the Bloom Desktop Graph App. 

The files inside the archive that you can download from the download center contain the assets of the Bloom app if you want to host Bloom on your own webserver, as well as the Bloom Server Plugin (the .jar files). There's some more information about the Bloom Server Plugin in the docs [here](https://neo4j.com/docs/bloom-user-guide/current/bloom-installation/installation-activation/#installation-bloom-server) and further information on how to set up an own webserver with the Bloom assets [here](https://neo4j.com/docs/bloom-user-guide/current/bloom-installation/advanced-installation/#bloom-web-app-on-separate-server). However, both these setups require a paid license for Bloom and aren't covered under the developer license that comes with Desktop.

To still use the latest version of Bloom for free as a developer, I would recommend using Bloom in Aura: https://neo4j.com/docs/bloom-user-guide/current/bloom-installation/bloom-deployment-modes/#bloom-in-aura. The latest version of Bloom is always available on https://bloom.neo4j.io and you can provide a url to your Neo4j database. Alternatively, you can also use [Explore](https://neo4j.com/docs/aura/explore/introduction/), which is the Bloom equivalent application inside Neo4j's [Aura Console](https://neo4j.com/docs/aura/new-console/). You can set up a Neo4j instance on Aura [free of charge](https://neo4j.com/press-releases/announcing-neo4j-auradb-free/).

I hope this helps. Please let me know if you have any other questions or comments

**Comment by ghanique** [comment]
@ckanz Thank you.

Leaves one remaining question: would it be an idea to ship the most recent version of Bloom with the next version of Neo4j Desktop, rather than an old version of it?

**Comment by ckanz** [comment]
> Leaves one remaining question: would it be an idea to ship the most recent version of Bloom with the next version of Neo4j Desktop, rather than an old version of it?

@ghanique graph apps within Desktop are updated independently from the Neo4j Desktop application itself and later versions of Bloom aren't available as graph apps anymore. However, the latest version of Bloom is available on https://bloom.neo4j.io/ and you can connect it to your Aura or Neo4j Desktop database.

**Comment by ckanz** [comment]
Hi @ghanique,

happy to let you know we recently released Neo4j Desktop 2.0: https://neo4j.com/blog/news/desktop-2/
You will be able to access the most recent version of Bloom (now called "Explore") there. 

Please let me know if you have question or comments.

**Comment by ghanique** [comment]
😁
